### PR TITLE
[Parse][CodeCompletion] Completions for precedencegroup decls

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Completions for precedence group contexts and precedence groups in operator declarations.

Resolves [SR-8043](https://bugs.swift.org/browse/SR-8043).

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
cc @benlangmuir

**Comment by AnthonyLatsis** [comment]
@jrose-apple Have I tagged the correct people to review this?

**Comment by jrose-apple** [comment]
Yep, those two seem fine! They'll tell you if it should be anyone else, too.

**Comment by AnthonyLatsis** [comment]
@benlangmuir @nkcsgexi sorry to disturb, reminder.

File: include/swift/Syntax/TokenKinds.def [hunk 1]
```diff
@@ -77,6 +77,12 @@
 #define PAT_KEYWORD(kw) SWIFT_KEYWORD(kw)
 #endif
 
+/// CONTEXTUAL_KEYWORD(kw)
+///   Keywords that are endemic to specific contexts, i.e. precedence group decls.
+#ifndef CONTEXTUAL_KEYWORD
+#define CONTEXTUAL_KEYWORD(kw) SWIFT_KEYWORD(kw)
+#endif
```
Comment (nkcsgexi) [inline include/swift/Syntax/TokenKinds.def]: hmm, could you avoid changing `TokenKinds.def` file? We currently parse these contextual keywords as identifiers and we have logic to promote them from identifier to keywords. Changing their kinds here can have larger impact than being necessary. Also, we have other contextual keywords too, e.g. `get` and `set`; it's hard to be exhaustive.

File: lib/IDE/CodeCompletion.cpp [hunk 2]
```diff
@@ -4124,6 +4126,79 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+        bool includePrivate = CurrDeclContext->getParentSourceFile() == file;
+        for (auto &pair : file->PrecedenceGroups) {
+          PrecedenceGroupDecl *group;
+          if ((group = pair.second.getPointer()) &&
+              (pair.second.getInt() || includePrivate))
+            addPrecedenceGroup(group);
+        }
+        break;
+      }
+      default: return;
```
Comment (nkcsgexi) [inline lib/IDE/CodeCompletion.cpp]: Nit: Move `return` to another line.

File: lib/IDE/CodeCompletion.cpp [hunk 3]
```diff
@@ -4124,6 +4126,79 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+
+    CurrDeclContext->getParentSourceFile()->
+      forAllVisibleModules([&](ModuleDecl::ImportedModule import) {
+       for (auto fileUnit: import.second->getFiles())
+          collectPrecedenceGroups(fileUnit);
+    });
+  }
+
+  void getPrecedenceGroupCompletions(SyntaxKind SK) {
+#define AddKeyword(kw) addKeyword(getTokenText(tok::kw), None,    \
```
Comment (nkcsgexi) [inline lib/IDE/CodeCompletion.cpp]: Could you make it a closure or function call instead of a macro for debuggability?

File: utils/gyb_syntax_support/Token.py [hunk 4]
```diff
@@ -88,6 +88,10 @@ def __init__(self, name, text):
     Keyword('__FUNCTION__', '__FUNCTION__'),
     Keyword('__DSO_HANDLE__', '__DSO_HANDLE__'),
     Keyword('Wildcard', '_'),
+    Keyword('Associativity', 'associativity'),
+    Keyword('HigherThan', 'higherThan'),
+    Keyword('LowerThan', 'lowerThan'),
+    Keyword('Assignment', 'assignment'),
```
Comment (nkcsgexi) [inline utils/gyb_syntax_support/Token.py]: Similarly, avoid adding these new keywords for consistency.

**Review by nkcsgexi — COMMENTED** [review]
Great progress overall! Some comments inline :)

**Review by benlangmuir — CHANGES_REQUESTED** [review]

File: include/swift/Syntax/TokenKinds.def (hunk 1 above)
Comment (AnthonyLatsis) [reply include/swift/Syntax/TokenKinds.def]: I thought this would be a nice starting point to clean up all of the string literal checks and conversions from identifiers that contextual keywords are currently suffering. In addition, even the highlighting is broken on them, possible as a consequence. Will this cause `Keyword cannot be used as an identifier here` in unnecessary places?

File: lib/IDE/CodeCompletion.cpp (hunk 3 above)
Comment (AnthonyLatsis) [reply lib/IDE/CodeCompletion.cpp]: If twice as much space wasted is not a problem – of course.

File: lib/Parse/ParseDecl.cpp [hunk 5]
```diff
@@ -6496,9 +6509,9 @@ Parser::parseDeclOperatorImpl(SourceLoc OperatorLoc, Identifier Name,
                                           Name, NameLoc, colonLoc,
                                           precedenceGroupName,
                                           precedenceGroupNameLoc);
-  
+
   diagnoseOperatorFixityAttributes(*this, Attributes, res);
-  
+
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: I knew you would say so 😅

File: lib/IDE/CodeCompletion.cpp [hunk 6]
```diff
@@ -4124,6 +4126,79 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+        semanticContext, {});
+      builder.addTextChunk(PGD->getName().str());
+      builder.setAssociatedDecl(PGD);
+    };
+
+    auto collectPrecedenceGroups = [&](FileUnit *FU) {
+      switch (FU->getKind()) {
+      case FileUnitKind::SerializedAST: {
+        SmallVector<Decl *, 16> topLevelDecls;
+        cast<LoadedFile>(FU)->getTopLevelDecls(topLevelDecls);
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/IDE/CodeCompletion.cpp]: Sure. Roughly, where do we cache those?

File: lib/IDE/CodeCompletion.cpp [hunk 7]
```diff
@@ -2740,7 +2742,7 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
       addTypeAnnotation(Builder, EnumType);
   }
 
-  void addKeyword(StringRef Name, Type TypeAnnotation,
+  void addKeyword(StringRef Name, Optional<Type> TypeAnnotation = None,
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/IDE/CodeCompletion.cpp]: Right, so that's the convention. I wasn't sure it was OK.

File: lib/IDE/CodeCompletion.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (benlangmuir) [reply lib/IDE/CodeCompletion.cpp]: The cache is in `CodeCompletionCache` and you request results from it using `RequestedCachedModule`.  Since you'll need a way to only include precedence groups and not all symbols from the module, look at how we use the flag `OnlyType` inside that request; you could add an `OnlyPredenceGroups` that works similarly.

The requests are put in `Lookup.RequestedCachedResults` and then handled in `handleResultsAndModules` and `copyCodeCompletionResults`.

File: lib/IDE/CodeCompletion.cpp (hunk 3 above)
Comment (nkcsgexi) [reply lib/IDE/CodeCompletion.cpp]: Yeah, space should be fine. Readability and debuggability are more important.

File: include/swift/Syntax/TokenKinds.def (hunk 1 above)
Comment (nkcsgexi) [reply include/swift/Syntax/TokenKinds.def]: These contextual keywords can be used as regular Swift identifiers, i.e. the code snippet is entirely valid:
```
func foo() {
    let get = 1
    let associativity = 2
    let higherThan = 3
    ...
}
```
However, the identifiers here are keywords in some circumstances. We have two ways to handle this bifurcation: (1) lexing them as identifier and later promote them during parsing to keywords, or (2) lexing them as contextual keywords and later demote them during parsing to identifier. The current implementation is using (1), and the change you made is changing it to (2). I cannot see any apparent benefit of using one over the other. So, that's why I suggest we avoid changing the kinds in `TokenKinds.def` to keep things simple.

Comment (AnthonyLatsis) [reply include/swift/Syntax/TokenKinds.def]: Ah. I agree, (1) is the most convenient.

File: lib/Parse/ParseDecl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: A separate PR or a simply a commit will do?

[the remark this answers is not in the document]
Comment (benlangmuir) [reply lib/Parse/ParseDecl.cpp]: Commit is fine

**Comment by AnthonyLatsis** [comment]
@benlangmuir I've hooked everything up, but apparently I still have to include precedence groups to what's cached per module, because they don't appear despite all other top-level declarations present in the cache sink. I don't see where this is regulated in `CodeCompletionCache`, could you help me out?

**Comment by benlangmuir** [comment]
@AnthonyLatsis We probably need to handle `PrecedenceGroupDecl` in `CompletionLookup::foundDecl` when `Kind == LookupKind::ImportFromModule`.  The flow would be something like:
```
SimpleCachingCodeCompletionConsumer::handleResultsAndModules
-> lookupCodeCompletionResultsFromModule
  -> CompletionLookup::getVisibleDeclsOfModule
    -> Module::lookupVisibleDecls
      ...
      -> CompletionLookup::foundDecl
```

**Comment by AnthonyLatsis** [comment]
@benlangmuir This appears to be a larger problem. I just realized `PrecendenceGroupDecl` is not a `ValueDecl`. Meanwhile, `VisibleDeclConsumer::foundDecl` takes `ValueDecl` and thus not only `ModuleDecl::lookupVisibleDecls` out of other `lookupVisibleDecls` looks exclusively for value decls. They should of been named `lookupVisibleValueDecls`..

**Comment by benlangmuir** [comment]
@AnthonyLatsis Argh. Sorry for leading you astray. I doubt we want to change the visible decl consumer interface, so we'll have to separately lookup the PrecedenceGroupDecls when we fill the cache.  In the worst case we could walk all the Decls like you were doing originally - at least this way we only do it once to fill the cache. 

@jrose-apple is there a better way to find all the PrecedenceGroupDecls in a module?  Should we add a new API to build an array of these from the data in ModuleFile ?

**Comment by AnthonyLatsis** [comment]
We still need a way to feed them to a `VisibleDeclConsumer`. The only relatively neat ways I see of doing this are adding a virtual (not pure)`foundPrecedenceGroup(...)` and the one you rejected. This way we can implement it only where necessary. I still think it would be convenient to rename `foundDecl` to `foundValueDecl` to avoid confusion if we decide to leave it as-is.

File: lib/Serialization/ModuleFile.cpp [hunk 8]
```diff
@@ -2013,6 +2013,16 @@ void ModuleFile::getTopLevelDecls(SmallVectorImpl<Decl *> &results) {
   }
 }
 
+void ModuleFile::getPrecedenceGroups(
```
Comment (benlangmuir) [inline lib/Serialization/ModuleFile.cpp]: Please add a pretty stack trace to the beginning of the function like we do in e.g. `getTopLevelDecls` so that if deserialization crashes we see this information.
```
PrettyStackTraceModuleFile stackEntry(*this);
```

File: lib/IDE/CodeCompletion.cpp [hunk 9]
```diff
@@ -5769,6 +5866,16 @@ void swift::ide::copyCodeCompletionResults(CodeCompletionResultSink &targetSink,
...
     });
+  } else if (onlyPrecedenceGroups) {
+    std::copy_if(sourceSink.Results.begin(), sourceSink.Results.end(),
+                 std::back_inserter(targetSink.Results),
+                 [](CodeCompletionResult *R) -> bool {
+      if (R->getAssociatedDeclKind() ==
+          CodeCompletionDeclKind::PrecedenceGroup) {
+        return true;
+      }
+      return false;
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: This could be simplified to `return R->getAssociatedDeclKind() == CodeCompletionDeclKind::PrecedenceGroup`.

File: lib/Parse/ParseDecl.cpp [hunk 10]
```diff
@@ -6594,21 +6610,39 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
       // We want to continue parsing after this.
       invalid = true;
     }
-    attrKeywordLoc = consumeToken(tok::identifier);
+    attrKeywordLoc = consumeToken();
```
Comment (benlangmuir) [inline lib/Parse/ParseDecl.cpp]: Why did this change?

File: lib/Parse/ParseDecl.cpp [hunk 11]
```diff
@@ -6655,46 +6703,53 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
...
         return abortBody();
       }
+      skipPostfixCodeCompletionToken(*lastTokTextEnd);
       continue;
     }
 
-    bool isLowerThan = false;
-    if (attrName == "higherThan" ||
-        (isLowerThan = (attrName == "lowerThan"))) {
+    if (attrName == "higherThan" || attrName == "lowerThan") {
```
Comment (benlangmuir) [inline lib/Parse/ParseDecl.cpp]: Is there a reason to change this?  Moving this code adds an extra string comparison (admittedly it might not matter, but it doesn't seem related to this patch).

File: lib/Parse/ParseDecl.cpp [hunk 12]
```diff
@@ -6655,46 +6703,53 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
...
         if (!consumeIf(tok::comma))
           break;
       } while (true);
       SyntaxContext->collectNodesInPlace(SyntaxKind::PrecedenceGroupNameList);
       continue;
     }
-    
-    diagnose(Tok, diag::unknown_precedencegroup_attribute, attrName);
+
+    diagnose(Tok, diag::unknown_precedencegroup_attribute, Tok.getText());
```
Comment (benlangmuir) [inline lib/Parse/ParseDecl.cpp]: Why did this change from `attrName`?  It ought to be the same value, but using `attrName` is less likely to break if someone refactors this code.

File: lib/IDE/CodeCompletion.cpp [hunk 13]
```diff
@@ -4205,6 +4224,67 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+          continue;
+
+        auto file = cast<SourceFile>(FU);
+        bool includePrivate = CurrDeclContext->getParentSourceFile() == file;
+
+        for (auto &pair : file->PrecedenceGroups) {
+          PrecedenceGroupDecl *group;
+          if ((group = pair.second.getPointer()) &&
+              (pair.second.getInt() || includePrivate))
+            addPrecedenceGroupRef(group);
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: Could we call `getPrecedenceGroups()` to avoid duplicating this logic?

File: test/IDE/complete_precedencegroup.swift [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+// ASSOCIATIVITY: Keyword/None: none;   name=none
+// ASSOCIATIVITY: Keyword/None: left;   name=left
+// ASSOCIATIVITY: Keyword/None: right;  name=right
+
+// ASSIGNMENT: Begin completions, 2 items
+// ASSIGNMENT: Keyword[false]/None: false; name=false
+// ASSIGNMENT: Keyword[true]/None: true;   name=true
+
+// PRECEDENCE_GROUP: Begin completions
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: BitwiseShiftPrecedence; name=BitwiseShiftPrecedence
```
Comment (benlangmuir) [inline test/IDE/complete_precedencegroup.swift]: Instead of checking all the decls from the stdlib, I suggest just checking one or two (e.g. Assignment and Comparison).  We prefer not to rely too much on stdlib details in code-completion tests.

File: test/IDE/complete_precedencegroup.swift [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: TernaryPrecedence; name=TernaryPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: DefaultPrecedence; name=DefaultPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: LogicalDisjunctionPrecedence; name=LogicalDisjunctionPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: LogicalConjunctionPrecedence; name=LogicalConjunctionPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: CastingPrecedence; name=CastingPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: AdditionPrecedence; name=AdditionPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: MultiplicationPrecedence; name=MultiplicationPrecedence
+// PRECEDENCE_GROUP: Decl[PrecedenceGroup]/OtherModule[Swift]: FunctionArrowPrecedence; name=FunctionArrowPrecedence
+
+// PRECEDENCE_GROUP_CURR_FILE-DAG: Decl[PrecedenceGroup]/CurrModule: MyPrecedence1; name=MyPrecedence1
```
Comment (benlangmuir) [inline test/IDE/complete_precedencegroup.swift]: Why are the current file results separated?  Shouldn't you get all the precedence groups in the stdlib and current file every time?

And we should probably have a test for a precedence group in another source file.

**Review by benlangmuir — COMMENTED** [review]
I'm really happy with the overall change now and this seems close to ready.  Thanks for working through the caching change!

**Comment by AnthonyLatsis** [comment]
Hey! Sorry for the delay, let's get this finished.

File: lib/Parse/ParseDecl.cpp (hunk 10 above)
Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: I guess I changed it to the shortcut version because we only call this lambda after we make sure it's an identifier, but yes, it isn't related to this PR and it's safer with the check, so I'll revert that.

File: test/IDE/complete_precedencegroup.swift (hunk 15 above)
Comment (AnthonyLatsis) [reply test/IDE/complete_precedencegroup.swift]: This is because I was having trouble getting precedence groups that are declared *after* the current one, they didn't show up. I thought this was happening because the rest hadn't been parsed yet and that's ok, but it doesn't really make sense now. I'll let you know if the issue is still present.

> And we should probably have a test for a precedence group in another source file.

It's a good idea, but I'll need a hint on how to write a multifile test. Do you have to manually create a module for that?

Comment (benlangmuir) [reply test/IDE/complete_precedencegroup.swift]: > I was having trouble getting precedence groups that are declared after the current one, they simply didn't show up.

Okay, we should fix that.  If it turns out to be hard to fix I'm okay with taking this change without it though.

> I'll need a hint on how to write a multifile test. Do you have to manually create a module for that?

No need for another module, just another file in the current module.  You can add another file with a precedence group in the `test/IDE/Inputs` directory and then add that file to the command line invocation of swift-ide-test on the RUN line.  See `complete_multiple_files.swift` for an example of how to do that.

File: lib/Parse/ParseDecl.cpp [hunk 16]
```diff
@@ -6601,14 +6617,32 @@ Parser::parseDeclPrecedenceGroup(ParseDeclOptions flags,
...
+      consumeToken();
+      return true;
+    }
+    return false;
+  };
+
+  // Skips the CC token if it comes without spacing.
+  auto skipPostfixCodeCompletionToken = [&](char lastTokTextEnd) {
+    if (Tok.is(tok::code_complete) && lastTokTextEnd != ' ' &&
+        lastTokTextEnd != '\t' && lastTokTextEnd != '\n')
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: I don't think this works because `Tok.getText()` doesn't contain white spaces.
Usually, to determine any whitespaces precedes a token, `getEndOfPreviousLoc() != Tok.getLoc()` is sufficient.
Also, if we skip `code_complete` token, we should return `hasCodeCompletion()` result even if it's ignored because further parsing is meaningless.
So I suggest:
```cpp
auto skipPostfixCodeCompletionToken = [&] () -> Bool {
  if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc()) {
    conumeToken();
    return true;
  }
  return false;
}
...
if (skipPostfixCodeCompletionToken())
  return abortBody(/*hasCodeCompletion*/true);
```

Comment (rintaro) [reply lib/Parse/ParseDecl.cpp]: But is this really needed in the first place? Although 
```swift
precedencegroup MyPrecedence {
  associativity: right higherThan: AdditionPrecedence
}
```
is valid syntax, consecutive attributes in a single line is not a good style IMO.

Something like:
```cpp
while (Tok.isNot(tok::r_brace)) {
  if (Tok.is(tok::code_completion)) {
    // Suggest attribute names only if CC token is at the start of line.
    if (Tok.isAtStartOfLine() && CodeCompletion) {
      CodeCompletion->completePrecedenceGroup(...)
    }
    consumeToken(tok::code_complete);
    continue;
  }
  ...
```
without `skipPostfixCodeCompletionToken()` is sufficient.

File: test/IDE/complete_precedencegroup.swift [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+precedencegroup MyPrecedence6 {
+  associativity: left
+  lowerThan: #^PRECEDENCE_GROUP_2^#
+  higherThan: AdditionPrecedence
+}
+precedencegroup MyPrecedence7 {
+  associativity: none
+  assignment: false
+  lowerThan: AdditionPrecedence
+  higherThan: #^PRECEDENCE_GROUP_3^#
```
Comment (rintaro) [inline test/IDE/complete_precedencegroup.swift]: Please add test case for `higherThan: SomePrecedence, #^TOKEN^#`

File: lib/Parse/ParseDecl.cpp (hunk 16 above)
Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: Placing the check before parsing the next attribute is a good move, though I think I've already tried and abandoned it for some reason. 

But, I believe that syntactic correctness is a superior criteria than 'bad style' for the compiler to support you.

> Usually, to determine any whitespaces precedes a token, `getEndOfPreviousLoc() != Tok.getLoc()` is sufficient. ... `if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc())`

Does `getEndOfPreviousLoc() != Tok.getLoc()` mean that there *is* or *isnt* a (`whitespace` | `\n` | `\r`)? Intuitively, the locations not matching tells me there is something in between, but not necessarily a symbol we are interested in.

Comment (rintaro) [reply lib/Parse/ParseDecl.cpp]: Between two adjacent tokens, only possible characters are whitespaces, newlines, comments, or other garbage characters *treated* as whitespaces (e.g. `U+00A0 No-Break Space`). Since comments are treated as a whitespace:
```swift
precedencegroup MyPrecedence {
  assignment: true/*comment*/higherThan: AssignmentPrecedence
}
```
is valid. So just checking (`whitespace` | `\n` | `\r`) is insufficient if we really want to check "whitespace"s.

Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: Right, I was actually just wondering if the second part of the condition from your reply had a typo (we should skip if there's, conversely, nothing in between: `getEndOfPreviousLoc() == Tok.getLoc()`)
```
if (Tok.is(tok::code_complete) && getEndOfPreviousLoc() != Tok.getLoc()) {
    conumeToken();
    return true;
}
```

Thanks for the tip on the preferred way of doing it.

Comment (rintaro) [reply lib/Parse/ParseDecl.cpp]: > should be `getEndOfPreviousLoc() == Tok.getLoc()`

Ah sorry. You are absolutely right!

**Comment by AnthonyLatsis** [comment]
`(higher | lower)Than: P1, ... , #^TOKEN^#` now works too, good catch @rintaro

**Comment by AnthonyLatsis** [comment]
@benlangmuir Finally got my hands on this. Fixed the delayed parsing issue.

**Comment by AnthonyLatsis** [comment]
Have a look at `Parser::consumeDecl` and `ContinueParsingAfterDelaying` to get a view of the fix. I basically use a flag to prevent skipping the whole file when consuming operator declarations that need a precedence group completion and precedence group declarations that need completion in a `PrecedenceRelation` attribute (precedence groups can be forward-referenced).

**Comment by AnthonyLatsis** [comment]
@rintaro or @nkcsgexi, mind taking a look in case Ben's busy?

File: lib/IDE/CodeCompletion.cpp [hunk 18]
```diff
@@ -4015,6 +4034,63 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+        RequestedResultsTy::fromModule(Module).onlyPrecedenceGroups());
+    });
+  }
+
+  void getPrecedenceGroupCompletions(SyntaxKind SK) {
+    switch (SK) {
+    case SyntaxKind::PrecedenceGroupAssociativity:
+      addKeyword("none");
+      addKeyword("left");
+      addKeyword("right");
```
Comment (nkcsgexi) [inline lib/IDE/CodeCompletion.cpp]: Could you refactor this mapping to a utility function like [here](https://github.com/apple/swift/blob/master/lib/AST/Attr.cpp#L38)? So other tooling can get the spelling as well. You can have a different commit to do that.

File: lib/IDE/CodeCompletion.cpp [hunk 19]
```diff
@@ -4015,6 +4034,63 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+      addKeyword(getTokenText(tok::kw_false), Type(), SemanticContextKind::None,
+                 CodeCompletionKeywordKind::kw_false);
+      addKeyword(getTokenText(tok::kw_true), Type(), SemanticContextKind::None,
+                 CodeCompletionKeywordKind::kw_true);
+      break;
+    case SyntaxKind::PrecedenceGroupAttributeList:
+      addKeyword("associativity");
+      addKeyword("higherThan");
+      addKeyword("lowerThan");
+      addKeyword("assignment");
```
Comment (nkcsgexi) [inline lib/IDE/CodeCompletion.cpp]: Same here.

File: lib/IDE/CodeCompletion.cpp [hunk 20]
```diff
@@ -1302,6 +1302,7 @@ class CodeCompletionCallbacksImpl : public CodeCompletionCallbacks {
   TypeLoc ParsedTypeLoc;
   DeclContext *CurDeclContext = nullptr;
   DeclAttrKind AttrKind;
+  SyntaxKind SyntxKind;
```
Comment (nkcsgexi) [inline lib/IDE/CodeCompletion.cpp]: It's a great idea to use syntax kind to differentiate completion scenario👍. However this is unprecedented. Could you add some doc-comments for what this SyntxKind should be used for?

**Comment by benlangmuir** [comment]
Sorry for the slow reply! I missed the last notification.  Thanks for tracking down what's going on with the forward reference case, this is actually a bigger issue than precedence groups!  The real problem here is that `consumeDecl` wants to use `IsTopLevel` to mean `TopLevelCode` - i.e. a script or main.swift file, where you cannot make forward references at all - but the caller is passing in `TopLevelLibrary` as well, where forward references are fine.  So we actually have the same bug already in e.g. inheritance completion where we only suggest the names earlier in the file.  This also means that that your change isn't quite right since a precendence group cannot use forward references inside a main.swift file, but your patch allows them.

Since this affects unrelated completions like inheritance, fixing it belongs in a separate PR and you can just ignore this bug for this one.

**Comment by AnthonyLatsis** [comment]
Ah, OK, I didn't have a broad enough picture of what's going on. I'll revert that and address the latest review comments. 

`TopLevelLibrary` is like a regular module, right?

**Comment by benlangmuir** [comment]
Yeah, it's just the top-level scope of the module.

File: lib/IDE/CodeCompletion.cpp (hunk 19 above)
Comment (AnthonyLatsis) [reply lib/IDE/CodeCompletion.cpp]: `Associativity` is pretty straight-forward, but we don't have an enum or something of the like for precedence group attributes, so I suppose there isn't any tooling depending on the spelling apart from individual cases that don't require switching on a specific PG attribute.

**Comment by AnthonyLatsis** [comment]
I believe all comments have been addressed, what do you think @everyone ?

**Review by benlangmuir — APPROVED** [review]
Thanks for working through it! Looks good to me.  Any remaining issues @rintaro and @nkcsgexi ?

**Comment by rintaro** [comment]
Conflicted with https://github.com/apple/swift/pull/19756.
Since that is experimental feature (for now), we don't need to care about the feature.
Could you simply resolve the conflict?

LGTM otherwise.

**Comment by AnthonyLatsis** [comment]
Resolved.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Thank you!

**Comment by AnthonyLatsis** [comment]
@rintaro, ever since I started contributing to the compiler I have been regularly receiving Swift-CI build failure notifications after merges. At first I thought it's an indication of me messing up things, but then I realized it could as well be other people who recently committed, or even something older. The tests that fail are often completely unfamiliar to me. This PR is not an exception, but I would like to ask whether there's anything else I should know about these notifications. I already know you can't reply to start a conversation :)

**Comment by rintaro** [comment]
Yeah, that is not your fault. Sorry for the noise.

> I already know you can't reply to start a conversation

See: https://forums.swift.org/c/development/ci-notifications
If I understand correctly, you reply to the email also appears in the forum.

**Comment by AnthonyLatsis** [comment]
Thanks! No worries about the noise, I don't mind getting them at all, just curious.
